ENH: Added several string distance functions

dtemkin opened the pull request:
Added Levenshtein, Damerau-Levenshtein, Hamming and Lee distance functions. I didn't see a tests module for distance metrics other than the demo at the bottom of the page. If someone knows where this located and is willing to point it out I would be more than happy to update the PR. Otherwise, I think it is all set to merged pending a review.

dtemkin committed:
Added string distance functions: Levenshtein Distance, Damerau-Levenshtein Distance, Hamming Distance, Lee Distance.

dtemkin committed:
Updated distance metrics documentation
Added Lee Distance Demo

alvations commented:
Thank you @dtemkin for the PR! This would be a good contribution to have in NLTK.

Actually there's already levenshtein distance in `nltk.metrics.distance.py`, it's the `edit_distance()` function. 

The implementations of the other distances looks clear but maybe there's some documentation steps that we could take to comment the code and like the previous PR on Jaro-Winkler similarity, some regression test would help maintainability of the code greatly. 

Let me try and work through the code review and see how we can improve so that NLTK users can benefit from the contribution =)

P/S: It's sort of conference season so it might take a while for me to get back to you regarding the reviews comments

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+
+    """
+    assert len(s1) == len(s2), "Strings must be the same length."
+    return sum(x[0] != x[1] for x in zip(s1, s2))
```
It's easier to read if you unpack the tuples during the zip comprehension, i.e. 

```
return sum(ch1 != ch2 for ch1, ch2 in zip(s1, s2))
```

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    Source(s):
+
+        (1) https://en.wikipedia.org/wiki/Hamming_distance
+
```
Maybe also cite 


> Hamming, R. W. (April 1950). "Error detecting and error correcting codes". The Bell System Technical Journal. 29 (2): 147–160.

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    n1, n2 = len(s1), len(s2)
+    sx = s1 + s2
+    for i in range(n1+n2):
+        alphabet.add(sx[i])
```
L499-503 could be simplified to 

```
alphabet = sorted(set(s1).union(set(s2)))
```

alvations commented on nltk/metrics/distance.py:
```diff
@@ -381,6 +535,15 @@ def demo():
+        print("Hamming Distance between '%s' and '%s':" % (s1, s2),
+              hamming_distance(s1, s2))
+        print("Lee Distance between '%s' and '%s':" % (s1, s2),
+              lee_distance(s1, s2))
```
A good test between Lee and Hamming distance would be to set a different `q` such that hamming distance is equivalent to Lee. This can be done by testing strings of length 2 and 3.

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+        for j in range(len(string)):
+            if alphabet[i] == string[j]:
+                string[j] = i
+    return string
```
Is this the expected output of the encoder?

```
>>> alphabet = ['1','2','3','4', '5', '6']
>>> s = list('12112345')
>>> _lee_string_encoder(s, alphabet)
[0, 1, 0, 0, 1, 2, 3, 4]
```

dtemkin replied:
Yes. Because it uses the index position of the character value as the encoded value.

alvations replied:
Could you add some description to the docstring to explain what the encoder is doing?

stevenbird replied:
@dtemkin would you mind taking care of this request, then I think we can merge this

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+
+    if q >= 2:
+        s1_enc = _lee_string_encoder(s1, alphabet=list(alphabet))
+        s2_enc = _lee_string_encoder(s2, alphabet=list(alphabet))
```
Casting a set into a list (depending on the version) could have different order, the distance value shouldn't change but the encoded string output changes. 

It's best to sort the set as shown in the above comment and avoid the list casting here.

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+        s2_enc = _lee_string_encoder(s2, alphabet=list(alphabet))
+        return sum([min([abs(s1_enc[i]-s2_enc[i]), q-abs(s1_enc[i]-s2_enc[i])]) for i in range(len(s1))])
+    else:
+        raise ValueError("number of distinct strings must be greater than 1")
```
try-except raise would be more natural than if-else raise

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+
+    q = len(alphabet)
+
+    if q >= 2:
```
Would be clearer with `if q > 1`, because the check is to make sure that the no. of distinct character is  >1.

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    if q >= 2:
+        s1_enc = _lee_string_encoder(s1, alphabet=list(alphabet))
+        s2_enc = _lee_string_encoder(s2, alphabet=list(alphabet))
+        return sum([min([abs(s1_enc[i]-s2_enc[i]), q-abs(s1_enc[i]-s2_enc[i])]) for i in range(len(s1))])
```
Similar to the above, we can avoid the index and retrieval by unpacking the string though zip since the checks on the character is always of the same index, i.e. 

```
return sum( min(abs(e1, e2), q - abs(e1, e2)) for e1, e2 in zip(s1_enc, s1_enc) )
```

alvations replied:
`sum()` should be able to take a generator so casting it into a list shouldn't be necessary =)

alvations commented:
I've left some comments on the Hamming and Lee distance since they're more straightforward. 
Will review the other 2 metrics later.

jnothman reviewed (commented):
Is Hamming distance relevant to natural language? Is Lee?

I wouldn't include those without clear justification from the NLP literature.

jnothman reviewed (commented):
Indeed Damereau-Levenshtein is already implemented in `edit_distance` which has a boolean `transpositions` parameter. `edit_distance` could have better documentation, but I fear this PR offers nothing of benefit to natural language processing beyond the existing provisions in NLTK.

alvations commented:
Hmmm, I'm curious now, whether Lee and Hamming distances are used NLP. 

Is it used in ispell / hunspell? Spell check systems? But then again, we could have used lev distance without the same length constraints.

jnothman commented:
I don't think Hamming or Lee are traditionally used in NLP, but I could be persuaded otherwise. They rely on positional correspondence between the strings being compared, which tends to be inappropriate for language (at the word or character level). I could only imagine them being used if, say, you had a morphological analysis of a word, such that you had a fixed set of slots you wanted to compare the values for. But even then, a generic implementation is not very interesting.

jnothman reviewed (commented):
Lee also seems to assume that the alphabet corresponds to ordinal values (or, rather, a group / rotational order) which is again very strange to apply to language. Giving users the impression that these are useful for comparing natural language strings is inappropriate, IMO.

53X commented:
Hamming distance is probably not directly related to NLP. Most of the others are already there in `nltk.metrics.distance.py`

dtemkin commented:
Hello all! thanks for taking the time to comment and review my PR. And a special thanks to @alvations and @jnothman for all the comments and code recommendations, I really appreciate it.

To address some of the comments/concerns, 
1) I didn't realize that the edit_distance accomplished the same objective as the Lev and Damerau-Lev although it makes sense now that I think about it. I will be sure to drop these on my next commit.

2) As far as the application of lee and hamming distance in NLP are concerned I would argue that since both algorithms are used, in part, to encode and correct messages sent over telecommunications networks they still hold some utility. One could also argue that they could be useful in identifying differences in genetic sequences which are, traditionally, presented as strings. 

However, I can see that, useful or not, the Lee and Hamming algorithms may not exactly fit in the nltk.metrics.distance.py module. But, what if they were in a distinct module for sequencing and/or differencing? Another benefit being it could provide a place for algorithms like the LCS, which I was surprised I didn't see already.  

Anyway, thanks for voicing your concerns, while I work on the next commit be sure to let me know your thoughts on adding a 'difference' module. 

As an aside, not long ago I was trying to construct an algorithm to classify abbreviated text. I could not get a good result due to the variation between strings and lack of shared convention among the different sources. But my model started performing alot better when I used a fully specified Huffman encoding scheme to preprocess the text. Point being, that you never know when something might become relevant until you hit that particular wall.

dtemkin committed:
Added source to Hamming Distance docs
Modified Lee and Hamming Distance Demos
Made requested improvements to code
Removed Levenshtein and Damerau-Levenshtein Distance metrics

dtemkin commented:
Thought of another possible implementation, as a 'second pass' for stop word filtering. Often, I find that I am excluding a lot of acronyms and abbreviations because my stop word list is never totally comprehensive and I end up filtering any word with a length that is less than some value.

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,10 +366,74 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    :param s1, s2: The strings to be analysed
+    :type s1: str
+    :type s2: str
+    :rtype int
```
missing colon:

```
:rtype: int
```

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,10 +366,74 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    :param s1, s2: The strings to be analysed
+    :type s1: str
+    :type s2: str
+    :rtype int
```
missing colon:

```
:rtype: int
```

alvations commented on nltk/metrics/distance.py:
```diff
@@ -366,10 +366,74 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    assert q > 1, "number of distinct characters must be greater than 1"
+
+    s1_enc = _lee_string_encoder(s1, alphabet=list(alphabet))
+    s2_enc = _lee_string_encoder(s2, alphabet=list(alphabet))
```
Casting to list is not necessary here, `sorted()` returns a list. https://docs.python.org/3/library/functions.html#sorted

dtemkin committed:
A couple minor syntax changes

jnothman reviewed (commented):
I still don't see how this is useful for language processing

stevenbird commented:
Closing; this is not sufficiently relevant to NLTK.
